tools/psa/build_tfm.py

```diff
@@ -0,0 +1,143 @@
+import sys
+import subprocess
+import logging
+
+logger = logging.getLogger('TF-M-Builder')
+logging.basicConfig(level=logging.INFO,
+                    format='[%(name)s] %(asctime)s: %(message)s.',
+                    datefmt='%H:%M:%S')
+
+ROOT = abspath(join(dirname(__file__), os.pardir, os.pardir))
+sys.path.insert(0, ROOT)
+
+TF_M_BUILD_DIR = join(ROOT, os.pardir, 'tfm_build_dir')
+VERSION_FILE_PATH = join(ROOT, 'features/FEATURE_PSA/FEATURE_TFM')
+
+dependencies = {
+    "trusted-firmware-m":
+    ['https://git.trustedfirmware.org/trusted-firmware-m.git',
+     'feature-twincpu'],
+    "mbedtls": ['https://github.com/ARMmbed/mbedtls.git',
+                'mbedtls-2.7.9'],
+    "mbed-crypto": ['https://github.com/ARMmbed/mbed-crypto.git',
+                    'mbedcrypto-1.1.0'],
+    "CMSIS_5": ['https://github.com/ARM-software/CMSIS_5.git', '5.5.0'],
+}
+
+def run_cmd_and_return_output(command):
+    """
+    Run the command in the sytem and return output.
+    Commands are passed as a list of tokens.
+    E.g. The command 'git remote -v' would be passed in as:
+     ['git', 'remote', '-v']
+
+    :param command: System command as a list of tokens
+    """
+    output = ''
+    logger.debug('[Exec] %s', ' '.join(command))
+    try:
+        fnull = open(os.devnull, 'w')
+        output = subprocess.check_output(command, stderr=fnull)
+    except subprocess.CalledProcessError as e:
+        logger.error("The command %s failed with return code: %s",
+                     (' '.join(command)), e.returncode)
+        sys.exit(1)
+    return output
+
+def detect_and_write_tfm_version(tfm_dir):
+    """
+    Identify the version of TF-M and write it to VERSION.txt
+    :param tfm_dir: The filesystem path where TF-M repo is cloned
+    """
+    cmd = ['git', '-C', tfm_dir, 'describe', '--tags',
+           '--abbrev=12', '--dirty', '--always']
+    tfm_version = run_cmd_and_return_output(cmd)
+    logger.info('TF-M version: %s', tfm_version)
+    if not isdir(VERSION_FILE_PATH):
+        os.makedirs(VERSION_FILE_PATH)
+    with open(join(VERSION_FILE_PATH, 'VERSION.txt'), 'w+') as f:
+        f.write(tfm_version)
+
+def check_repo_version(name, deps):
+    """
+    Compare the version of cloned and expected and exit if they dont match
+    :param name: Name of the git repository
+    :param deps: Dictionary containing dependency details
+    """
+    basedir = TF_M_BUILD_DIR
+    if name == 'trusted-firmware-m':
+        cmd = ['git', '-C', join(basedir, name),
+               'rev-parse', '--abbrev-ref', 'HEAD']
+    else:
+        cmd = ['git', '-C', join(basedir, name),
+               'describe', '--tags']
+    _out = run_cmd_and_return_output(cmd)
+    if _out.strip('\n') != deps.get(name)[1]:
+        logger.error('Conflict: cloned "%s" and expected "%s"',
+                     _out.strip('\n'), deps.get(name)[1])
+        logger.error('check and remove folder %s',
+                     join(basedir, name))
+        sys.exit(1)
```

Comment:
Do not call sys.exit(1). This will crash out of the program and will do no cleanup. It also will fail in the online system.

Reply:
@mark-edgeworth  In this case we need to exit from the script without doing any cleanup since the cloned repos might contain user changes. Therefore, the only thing we do here is to warn the user and let them take the action. 
Having said that, this path will never be triggered in the online system as there won't be any changes done by it.